Foundation/FileManager+Win32.swift

```diff
@@ -648,7 +644,95 @@ extension FileManager {
     }
 
     internal func _contentsEqual(atPath path1: String, andPath path2: String) -> Bool {
-        NSUnimplemented()
+        guard let path1Handle = path1.withCString(encodedAs: UTF16.self, {
+            CreateFileW($0, GENERIC_READ, DWORD(FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE), nil,
+                        DWORD(OPEN_EXISTING), DWORD(FILE_FLAG_OPEN_REPARSE_POINT | FILE_FLAG_BACKUP_SEMANTICS), nil)
+        }) else {
+            return false
+        }
+        defer { CloseHandle(path1Handle) }
+
+        guard let path2Handle = path2.withCString(encodedAs: UTF16.self, {
+            CreateFileW($0, GENERIC_READ, DWORD(FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE), nil,
+                        DWORD(OPEN_EXISTING), DWORD(FILE_FLAG_OPEN_REPARSE_POINT | FILE_FLAG_BACKUP_SEMANTICS), nil)
+        }) else {
+            return false
+        }
+        defer { CloseHandle(path2Handle) }
+
+        guard path1Handle != INVALID_HANDLE_VALUE, path2Handle != INVALID_HANDLE_VALUE else {
+            return false
+        }
+
+        let file1Type = GetFileType(path1Handle)
+        guard GetLastError() == NO_ERROR else {
+            return false
+        }
+        let file2Type = GetFileType(path2Handle)
+        guard GetLastError() == NO_ERROR else {
+            return false
+        }
+
+        guard file1Type == file2Type, file1Type == FILE_TYPE_DISK else {
+            return false
+        }
+
+        var path1FileInfo = BY_HANDLE_FILE_INFORMATION()
+        var path2FileInfo = BY_HANDLE_FILE_INFORMATION()
+        guard GetFileInformationByHandle(path1Handle, &path1FileInfo),
+              GetFileInformationByHandle(path2Handle, &path2FileInfo) else {
+            return false
+        }
+
+        // If both paths point to the same volume/filenumber or they are both zero length
+        // then they are considered equal
+        let sameFile = path1FileInfo.nFileIndexHigh == path2FileInfo.nFileIndexHigh
+              && path1FileInfo.nFileIndexLow == path2FileInfo.nFileIndexLow
+              && path1FileInfo.dwVolumeSerialNumber == path2FileInfo.dwVolumeSerialNumber
+
+        if sameFile {
+            return true
+        }
+
+        let path1Attrs = path1FileInfo.dwFileAttributes
+        let path2Attrs = path2FileInfo.dwFileAttributes
+        if DWORD(FILE_ATTRIBUTE_REPARSE_POINT) & path1Attrs == DWORD(FILE_ATTRIBUTE_REPARSE_POINT)
+             || DWORD(FILE_ATTRIBUTE_REPARSE_POINT) & path2Attrs == FILE_ATTRIBUTE_REPARSE_POINT {
+            guard DWORD(FILE_ATTRIBUTE_REPARSE_POINT) & path1Attrs == DWORD(FILE_ATTRIBUTE_REPARSE_POINT)
+                    && DWORD(FILE_ATTRIBUTE_REPARSE_POINT) & path2Attrs == FILE_ATTRIBUTE_REPARSE_POINT else {
+                return false
+            }
```

Comment:
Huh?  What am I missing?  Isn't this the condition that is in the `if`?

Reply:
The inner one is `&&`, the outer is `||`. If either is a reparse point, go into the `if`, but if one is not a reparse point, the immediately bail out. Sadly there's no logical XOR which might make that clearer.

Reply:
Hmm, does `^` not work here?